File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists.length == 0){
            return null;
        }
        ListNode arr = mergeSort(lists, 0, lists.length - 1);
        return arr;
    }
    private ListNode mergeSort(ListNode[] lists, int start, int end){
        if(start == end){
            return lists[end];
        }

        int mid = (end + start) / 2;
        ListNode left = mergeSort(lists, start, mid);
        ListNode right = mergeSort(lists, mid + 1, end);
        ListNode arr = merge(right, left);
        return arr;
    }
    private ListNode merge(ListNode right, ListNode left){
        ListNode merged = new ListNode(-1);
        ListNode head = merged;
        if(right != null && left != null){
            while(right != null && left != null){
                if(right.val <= left.val){
                    merged.next = right;
                    ListNode temp = right.next;
                    right.next = null;
                    right = temp;
                }
                else{
                    merged.next = left;
                    ListNode temp = left.next;
                    left.next = null;
                    left = temp;
                }
                merged = merged.next;
            }
            if(right == null){
                merged.next = left;
            }
            if(left == null){
                merged.next = right;
            }
        }
        else if(right == null){
            return left;
        }
        else if(left == null){
            return right;
        }
        return head.next;
    }
}

File: Data Structures & Algorithms/mergeSort/submission-0.cs
// Definition for a pair.
// public class Pair {
//     public int Key;
//     public string Value;
//
//     public Pair(int key, string value) {
//         Key = key;
//         Value = value;
//     }
// }
public class Solution {
    public List<Pair> MergeSort(List<Pair> pairs) {
         return Sort(pairs, 0, pairs.Count - 1);
    }
    public List<Pair> Sort(List<Pair> pairs, int st, int end){
        if(end - st + 1 <= 1){
            return pairs;
        }
        int mid = (st + end) / 2;
        Sort(pairs, st, mid);
        Sort(pairs, mid + 1, end);
        Merge(pairs, st, mid, end);
        return pairs;
    }
    private void Merge(List<Pair> pairs, int st, int mid, int end){
        int i = 0, j = 0, k = st;
        List<Pair> L = new List<Pair>(pairs.GetRange(st, mid - st + 1));
        List<Pair> R = new List<Pair>(pairs.GetRange(mid + 1, end - mid));
        while (i < L.Count && j < R.Count){
            if(L[i].Key <= R[j].Key){
                pairs[k] = L[i];
                i++;
            }
            else{
                pairs[k] = R[j];
                j++;
            }
            k++;
        }
        while (i < L.Count) {
            pairs[k] = L[i];
            i++;
            k++;
        }
        while (j < R.Count) {
            pairs[k] = R[j];
            j++;
            k++;
        }
    }
}
